core/src/main/java/de/jplag/SubmissionSet.java

```diff
@@ -152,30 +167,58 @@ private void parseBaseCodeSubmission(Submission baseCode) throws BasecodeExcepti
      * Parse all given submissions.
      * @param submissions The list of submissions
      */
-    private void parseSubmissions(List<Submission> submissions) throws LanguageException {
+    private void parseSubmissions(List<Submission> submissions) throws ExitException {
         if (submissions.isEmpty()) {
             logger.error("No submissions to parse!");
             return;
         }
 
         ProgressBar progressBar = ProgressBarLogger.createProgressBar(ProgressBarType.PARSING, submissions.size());
-        for (Submission submission : submissions) {
-
-            logger.trace("------ Parsing submission: " + submission.getName());
-            currentSubmissionName = submission.getName();
 
-            boolean successful = submission.parse(options.debugParser(), options.normalize(), options.minimumTokenMatch());
-            if (!successful) {
-                errors++;
-                logger.debug("ERROR -> Submission {} removed with reason {}", currentSubmissionName, submission.getState());
+        if (options.language().expectsSubmissionOrder()) {
+            for (Submission submission : submissions) {
+                parseSingleSubmission(progressBar, submission);
             }
-            progressBar.step();
+        } else {
+            parseSubmissionsInParallel(submissions, progressBar);
         }
+
         progressBar.dispose();
 
-        int validSubmissions = submissions.size() - errors;
+        int validSubmissions = submissions.size() - errors.get();
         logger.debug("{} submissions parsed successfully!", validSubmissions);
-        logger.debug("{} parser error{}!", errors, errors != 1 ? "s" : "");
+        logger.debug("{} parser error{}!", errors, errors.get() != 1 ? "s" : "");
+    }
+
+    private void parseSubmissionsInParallel(List<Submission> submissions, ProgressBar progressBar) throws SubmissionException {
+        try (ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor()) {
+            List<Callable<Void>> tasks = new ArrayList<>();
+            for (Submission submission : submissions) {
+                tasks.add(() -> {
+                    parseSingleSubmission(progressBar, submission);
+                    return null; // Ensure the lambda is a Callable
+                });
+            }
+            for (Future<Void> future : executor.invokeAll(tasks)) {
+                future.get();
```

Comment:
Why do you need the get & iterate. Invoke all + await termination could do the trick, right?